### PR TITLE
Separate fetching and parsing of merchandising feeds

### PR DESCRIPTION
**Pull request by kelvin-chappell**
This PR is just for jobs and soulmates feeds.

It has several benefits:
- Can use local copy of stale data if it's not possible to fetch a feed
- Have copy of bad data if it's not possible to parse the feed
- Have separate metrics for fetching and parsing
- Can continue to parse last known good data if current feed is bad

Still to do:
- Other feeds (masterclasses, books, travel, best buys etc)
- storing last good version of data separately from last data fetched
- Set alarms based on parsing metrics

### CODE REVIEW

File: commercial/app/commercial/CommercialLifecycle.scala [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+      def storeFeed(feed: Feed): Unit =
+        S3.putPrivate(key = s"$merchandisingFeedsRoot/$feedName", value = feed.content, feed.contentType)
+
+      def recordFetch(maybeDuration: Option[Duration]): Unit = {
+        recordEvent(feedName, "fetch", maybeDuration)
+      }
+
+      val msgPrefix = s"Fetching $feedName feed"
+      log.info(s"Fetching $feedName feed from ${fetcher.url} ...")
```
Comment (kenlim) [inline commercial/app/commercial/CommercialLifecycle.scala]: Maybe you should just reuse msgPrefix

File: commercial/app/commercial/CommercialLifecycle.scala [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+    val key = s"${feedName.toLowerCase.replaceAll("[\\s/]+", "-")}-$eventName-time"
+    val duration = maybeDuration map (_.toMillis.toDouble) getOrElse -1d
+    CloudWatch.put("Commercial", Map(s"$key" -> duration))
+  }
+
+  override def onStart(app: PlayApp): Unit = {
+
+    val randomFactor = Random.nextInt(15)
+
+    def randomStartSchedule(minsLater: Int = 0) = s"0 ${randomFactor + minsLater}/15 * * * ?"
```
Comment (kenlim) [inline commercial/app/commercial/CommercialLifecycle.scala]: Maybe move the randomFactor val right in here.

File: commercial/app/commercial/feeds/FeedParser.scala [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    for (agent <- SoulmatesAgent.agents) yield {
+      new FeedParser[Member] {
+        val feedName = s"soulmates/${agent.groupName}"
+
+        def parse(): Future[ParsedFeed[Member]] = agent.refresh()
+      }
+    }
+  }
+
+  val all = Seq(jobs) ++ soulmates
```
Comment (kenlim) [inline commercial/app/commercial/feeds/FeedParser.scala]: Why is the signature between jobs and soulmates so different?

File: common/app/conf/switches/Switches.scala [hunk 4]
```diff
@@ -46,6 +46,9 @@ trait SwitchTrait extends Switchable with Initializable[SwitchTrait] {
 
   def isSwitchedOn: Boolean = delegate.isSwitchedOn && new LocalDate().isBefore(sellByDate)
 
+  // true if switched on and switchboard has been read
```
Comment (kenlim) [inline common/app/conf/switches/Switches.scala]: Change to "If the switchboard hasn't been properly initialised, then the "safe state" is returned, not the real switch value. This makes sure the switchboard is initialised before returning the answer"

File: commercial/app/model/commercial/jobs/JobsFeed.scala [hunk 5]
```diff
@@ -1,52 +1,42 @@
...
-      parse(XML.loadString(body.dropWhile(_ != '<')))
-    }
-    } getOrElse {
-      log.warn(s"Missing URL for Jobs feed")
-      Future.failed(FeedMissingConfigurationException("Jobs"))
+  def parsedJobs(feedName: String): Future[ParsedFeed[Job]] = {
+    Switches.JobFeedSwitch.isGuaranteedSwitchedOn flatMap { switchedOn =>
+      if (switchedOn) {
+        val start = currentTimeMillis
+        S3.get(s"$merchandisingFeedsRoot/$feedName") map { body =>
```
Comment (kenlim) [inline commercial/app/model/commercial/jobs/JobsFeed.scala]: Should wrap this in a very obvious function that this is the "persistent" object that is giving us the safety.

File: commercial/app/commercial/CommercialLifecycle.scala [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+
+    def fetchFeed(fetcher: FeedFetcher): Unit = {
+
+      val feedName = fetcher.feedName
+
+      def storeFeed(feed: Feed): Unit =
+        S3.putPrivate(key = s"$merchandisingFeedsRoot/$feedName", value = feed.content, feed.contentType)
```
Comment (kenlim) [inline commercial/app/commercial/CommercialLifecycle.scala]: This is persisting our precious object. Should wrap into a function with obvious name that this is special.

File: commercial/app/commercial/feeds/FeedParser.scala (hunk 3 above)
Comment (kelvin-chappell) [inline commercial/app/commercial/feeds/FeedParser.scala]: There are many soulmates feeds and only one jobs feed.

**Pull request merged**